components/eamxx/src/physics/mam/readfiles/fractional_land_use_impl.hpp

```diff
@@ -0,0 +1,164 @@
+                     "ERROR: fractional land use data is on a different grid "
+                     "than the model one,\n"
+                     "       but remap file is missing from fractional "
+                     "land use parameter list.");
+
+    remapper =
+        std::make_shared<RefiningRemapperP2P>(horiz_interp_tgt_grid, map_file);
+  }
+
+  remapper->registration_begins();
+
+  const auto tgt_grid = remapper->get_tgt_grid();
+
+  const auto layout_2d = tgt_grid->get_2d_vector_layout(nclass_data, "class");
+  const auto nondim    = ekat::units::Units::nondimensional();
+
+  Field fractional_land_use(
+      FieldIdentifier(field_name, layout_2d, nondim, tgt_grid->name()));
+  fractional_land_use.allocate_view();
+
+  remapper->register_field_from_tgt(fractional_land_use);
+
+  remapper->registration_ends();
+
+  return remapper;
+}  // create_horiz_remapper
+
+// -------------------------------------------------------------------------------------------
+// -------------------------------------------------------------------------------------------
+
+template <typename S, typename D>
+std::shared_ptr<AtmosphereInput> fracLandUseFunctions<S, D>::create_data_reader(
+    const std::shared_ptr<AbstractRemapper> &horiz_remapper,
+    const std::string &data_file) {
+  std::vector<Field> io_fields;
+  for(int i = 0; i < horiz_remapper->get_num_fields(); ++i) {
+    io_fields.push_back(horiz_remapper->get_src_field(i));
+  }
+  const auto io_grid = horiz_remapper->get_src_grid();
+  return std::make_shared<AtmosphereInput>(data_file, io_grid, io_fields, true);
+}  // create_data_reader
+
+// -------------------------------------------------------------------------------------------
+// -------------------------------------------------------------------------------------------
+
+template <typename S, typename D>
+void fracLandUseFunctions<S, D>::update_frac_land_use_data_from_file(
+    std::shared_ptr<AtmosphereInput> &scorpio_reader,
+    AbstractRemapper &horiz_interp, FracLandUseInput &input) {
+  using namespace ShortFieldTagsNames;
+  using ESU    = ekat::ExeSpaceUtils<typename DefaultDevice::execution_space>;
+  using Member = typename KokkosTypes<DefaultDevice>::MemberType;
+
+  start_timer("EAMxx::FracLandUse::update_frac_land_use_data_from_file");
+
+  // 1. Read from file
+  start_timer(
+      "EAMxx::FracLandUse::update_frac_land_use_data_from_file::read_data");
+  scorpio_reader->read_variables();
+  stop_timer(
+      "EAMxx::FracLandUse::update_frac_land_use_data_from_file::read_data");
+
+  // 2. Run the horiz remapper (it is a do-nothing op if FracLandUse data is on
+  // same grid as model)
+  start_timer(
+      "EAMxx::FracLandUse::update_frac_land_use_data_from_file::horiz_remap");
+  horiz_interp.remap(/*forward = */ true);
+  stop_timer(
+      "EAMxx::FracLandUse::update_frac_land_use_data_from_file::horiz_remap");
+
+  // 3. Copy from the tgt field of the remapper into the FracLandUse_data,
```

Comment:
The pad in SPA was needed b/c of vertical interpolation. Here, you don't have it, and in fact you are simply calling a deep_copy> I think this should be cleaned up. You don't need a deep copy, you can make input.data use the exact same view that is in the tgt field.

Reply:
I have implemented it they way you recommended and it is much cleaner. Thanks!